src/worker/vm/extensions/fetch.ts

```diff
@@ -0,0 +1,80 @@
+import dns from 'dns'
+import ipRangeCheck from 'ip-range-check'
+import net from 'net'
+import nodeFetch, { HeaderInit, RequestInit, Response } from 'node-fetch'
+import url from 'url'
+
+import { IllegalOperationError } from '../../plugins/run'
+
+export interface URLLike {
+    href: string
+}
+
+// ['93.184.216.34', '172.31.0.0/16']
+const RESTRICTED_IP_RANGES =
+    process.env.NODE_ENV === 'test'
+        ? ['127.0.0.1/8', '116.203.255.68']
+        : (process.env.RESTRICTED_IP_RANGES || '').split(',')
+
+const isIpInRestrictedRange = (ip: string): boolean => {
+    return ipRangeCheck(ip, RESTRICTED_IP_RANGES)
+}
+
+const validateHostOrUrl = async (hostOrUrl: any) => {
+    if (typeof hostOrUrl !== 'string') {
+        throw new IllegalOperationError(`Invalid host/URL ${hostOrUrl}: Not a string or URLLike object`)
+    }
+
+    if (net.isIP(hostOrUrl) && isIpInRestrictedRange(hostOrUrl)) {
+        throw new IllegalOperationError(`IP ${hostOrUrl} is not allowed for security reasons`)
+    }
+
+    if (hostOrUrl.startsWith('http://localhost')) {
+        throw new IllegalOperationError(`${hostOrUrl} is not allowed for security reasons`)
+    }
+
+    const parsedHost = url.parse(hostOrUrl).hostname
+    if (!parsedHost) {
+        throw new IllegalOperationError(`Could not find a hostname for ${hostOrUrl}`)
+    }
+
+    const lookupResult = await dns.promises.lookup(parsedHost)
+    if (isIpInRestrictedRange(lookupResult.address)) {
+        throw new IllegalOperationError(`IP ${lookupResult.address} is not allowed for security reasons`)
+    }
+}
+
+export const fetch = async (url: string | URLLike, init?: RequestInit): Promise<Response> => {
+    if (typeof url === 'object' && 'href' in url) {
+        url = url.href
+    }
+    await checkRedirectChain(url, init?.headers)
```

Comment:
I'm worried that this could result in bad issues if we ping a non-idempotent URL more than once this way (and discarding the result too). I think it'd be better to forgo this check.